### ISSUE TITLE
"openssl sha1 -verify" fails with OpenSSL 3.0.7

### ISSUE THREAD

**Issue by mkauf** [openssl/openssl#19589]
There is a regression in OpenSSL 3.0.7 in the "openssl sha1" command.

To test this, please perform these steps with OpenSSL 3.0.5:
```
echo "test" > test.dat
openssl genrsa -out key.pem 4096
openssl rsa -in key.pem -outform PEM -pubout -out public.pem
openssl sha1 -sign key.pem -out signature.dat test.dat
```

Perform this step with OpenSSL 3.0.7:
```
openssl sha1 -verify public.pem -signature signature.dat test.dat
```

Actual result:
```
Verification failure
40E7076C4C7F0000:error:02000068:rsa routines:ossl_rsa_verify:bad signature:crypto/rsa/rsa_sign.c:430:
40E7076C4C7F0000:error:1C880004:Provider routines:rsa_verify:RSA lib:providers/implementations/signature/rsa_sig.c:774:
```

Expected result:
```
Verified OK
```

Note: It works if the command "openssl dgst" is used instead:
```
openssl dgst -sha1 -verify public.pem -signature signature.dat test.dat
```

The test can also be performed with "opposite" version numbers (generate the signature with OpenSSL 3.0.7 and verify it with OpenSSL 3.0.5), this also doesn't work.

Tested on Linux (x86_64), with GCC 8.5 and GCC 11.

**Comment by mkauf** [comment]
"git bisect" shows that this is the bad commit: 4fe6c2fdc80b11d6ec16c7a87ec46031d459e86b

CC @neverpanic , @beldmit

I think the bug is that the digest algorithm is selected here:
```
md = (EVP_MD *)EVP_get_digestbyname(argv[0]);
```

... but if `impl == NULL`, the variable `md` is not used, and `digestname` is used instead:
```
           if (impl == NULL)
                res = EVP_DigestVerifyInit_ex(mctx, &pctx, digestname,
                                              app_get0_libctx(),
                                              app_get0_propq(), sigkey, NULL);
            else
                res = EVP_DigestVerifyInit(mctx, &pctx, md, impl, sigkey);
```

**Comment by t8m** [comment]
Fix in #19606

**Comment by t8m** [comment]
Thank you for the report and bisect!

**Comment by joelharkes** [comment]
we are recently getting this error in our PHP SQL Server driver.

```
[Microsoft][ODBC Driver 18 for SQL Server]SSL Provider: [error:02000068:rsa routines::bad signature][error:1C880004:Provider routines::RSA lib] (Connection: sqlsrv, SQL: update 
```

**Comment by joelharkes** [comment]
i guess upgrading to 3.2.x should do the fix?
https://github.com/openssl/openssl/commit/1e5780dbc79dab14c1ec1584313755fc2fd2cf55

**Comment by mkauf** [comment]
@joelharkes The bug has been fixed in OpenSSL 3.0.8.

The new problem that you describe is probably unrelated.
